### PR TITLE
Add support for Detection task type

### PR DESCRIPTION
**Pull request by djdameln**
# Description

This PR adds the detection task type. This is mainly to support region-based models that produce bounding box predictions instead of / in addition to anomaly maps.

### Major
- Adds conversion utilities between masks and bounding boxes. This allows automatic generation of ground truth boxes for detection models. These utilities are also used to generate bounding box predictions when the model is of the segmentation type.
- The detection task uses the thresholding and evaluation mechanisms from the segmentation task under the hood. This is achieved by converting the bounding boxes to segmentation masks or anomaly maps before evaluating the metrics.
- Visualization functionality was added. This means that the predictions of any model capable of localization (segmentation or detection) can be visualized as detection predictions. For example, the following image was generated by passing `detection` as task type for a padim model training on MVTec bottle dataset:

### Minor
- Using enum instead of string to configure task type.
- A new `collate_fn` function had to be implemented to allow collating the ground truth bounding boxes across batches.

## Known issues
- No test cases (WIP)
- No inference support

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/data/__init__.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 from .folder import Folder
 from .inference import InferenceDataset
 from .mvtec import MVTec
+from .task_type import TaskType
```
Comment (samet-akcay) [inline anomalib/data/__init__.py:17]: would this be used by models and other components as well? If so, would it be an idea to move it to a place that would be accessible by other components?

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for the efforts! I have a few minor comments

File: anomalib/data/base/datamodule.py [hunk 2]
```diff
@@ -12,14 +12,28 @@
...
 
 from anomalib.data.base.dataset import AnomalibDataset
 from anomalib.data.utils import ValSplitMode, random_split
 
 logger = logging.getLogger(__name__)
 
 
+def collate_fn(batch):
+    """Custom collate function that collates bounding boxes as lists."""
+    elem = batch[0]
```
Comment (samet-akcay) [inline anomalib/data/base/datamodule.py:25]: What does `elem` mean, element? Would it be possible to rename the variable to show the intention?
Also what else the type would be? Would it be possible to annotate it?

File: anomalib/data/base/datamodule.py [hunk 3]
```diff
@@ -101,8 +115,20 @@ def train_dataloader(self) -> TRAIN_DATALOADERS:
...
     def val_dataloader(self) -> EVAL_DATALOADERS:
         """Get validation dataloader."""
-        return DataLoader(self.val_data, shuffle=False, batch_size=self.eval_batch_size, num_workers=self.num_workers)
+        return DataLoader(
+            self.val_data,
+            shuffle=False,
+            batch_size=self.eval_batch_size,
+            num_workers=self.num_workers,
+            collate_fn=collate_fn,
+        )
```
Comment (samet-akcay) [inline anomalib/data/base/datamodule.py:124]: Sorry, it's a bit picky comment, but would be be easier to read if all the arguments are specified.
```suggestion
        return DataLoader(
            dataset=self.val_data,
            shuffle=False,
            batch_size=self.eval_batch_size,
            num_workers=self.num_workers,
            collate_fn=collate_fn,
        )
```

File: anomalib/data/base/datamodule.py [hunk 4]
```diff
@@ -101,8 +115,20 @@ def train_dataloader(self) -> TRAIN_DATALOADERS:
...
     def test_dataloader(self) -> EVAL_DATALOADERS:
         """Get test dataloader."""
-        return DataLoader(self.test_data, shuffle=False, batch_size=self.eval_batch_size, num_workers=self.num_workers)
+        return DataLoader(
+            self.test_data,
+            shuffle=False,
+            batch_size=self.eval_batch_size,
+            num_workers=self.num_workers,
+            collate_fn=collate_fn,
+        )
```
Comment (samet-akcay) [inline anomalib/data/base/datamodule.py:134]:
```suggestion
        return DataLoader(
            dataset=self.test_data,
            shuffle=False,
            batch_size=self.eval_batch_size,
            num_workers=self.num_workers,
            collate_fn=collate_fn,
        )
```

File: anomalib/data/base/video.py [hunk 5]
```diff
@@ -74,9 +78,15 @@ def __getitem__(self, index: int) -> Dict[str, Union[str, Tensor]]:
...
             item["image"] = torch.stack([item["image"] for item in processed_frames]).squeeze(0)
-            mask = item["mask"]
+            mask = Tensor(item["mask"])
             item["mask"] = torch.stack([item["mask"] for item in processed_frames]).squeeze(0)
             item["label"] = Tensor([1 in frame for frame in mask]).int().squeeze(0)
+            if self.task == TaskType.DETECTION:
+                item["boxes"] = [
+                    torch.empty((0, 4)) if frame.max() == 0 else masks_to_boxes(frame)
+                    for frame in item["mask"].view((-1, 1) + item["mask"].shape[-2:])
+                ]
```
Comment (samet-akcay) [inline anomalib/data/base/video.py:88]: This line is a bit difficult to follow. Would it be possible to have a specific function for this line so the reader could go over it line by line for clarity?

File: anomalib/data/folder.py [hunk 6]
```diff
@@ -141,7 +142,7 @@ class FolderDataset(AnomalibDataset):
     """Folder dataset.
 
     Args:
-        task (str): Task type. (classification or segmentation).
+        task (TaskType): Task type. (classification, detection or segmentation).
```
Comment (samet-akcay) [inline anomalib/data/folder.py:145]: For consistency, following your changes above.
```suggestion
        task (TaskType): Task type. (``classification``, ``detection`` or ``segmentation``).
```

File: anomalib/data/folder.py [hunk 7]
```diff
@@ -222,8 +223,8 @@ class Folder(AnomalibDataModule):
         train_batch_size (int, optional): Training batch size. Defaults to 32.
         test_batch_size (int, optional): Test batch size. Defaults to 32.
         num_workers (int, optional): Number of workers. Defaults to 8.
-        task (str, optional): Task type. Could be either classification or segmentation.
-            Defaults to "classification".
+        task (TaskType, optional): Task type. Could be classification, detection or segmentation.
+            Defaults to segmentation.
```
Comment (samet-akcay) [inline anomalib/data/folder.py:227]:
```suggestion
        task (TaskType, optional): Task type. Could be ``classification``, ``detection`` or ``segmentation``.
            Defaults to ``segmentation``.
```

File: anomalib/data/mvtec.py [hunk 8]
```diff
@@ -123,7 +124,7 @@ class MVTecDataset(AnomalibDataset):
     """MVTec dataset class.
 
     Args:
-        task (str): Task type, either 'classification' or 'segmentation'
+        task (TaskType): Task type,'classification', 'detection' or 'segmentation'
```
Comment (samet-akcay) [inline anomalib/data/mvtec.py:127]:
```suggestion
        task (TaskType): Task type, ``classification``,``detection`` or ``segmentation``
```

File: anomalib/data/task_type.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
+"""Task type enum."""
+
```
Comment (samet-akcay) [inline anomalib/data/task_type.py:2]: Would be good to add the license.

File: anomalib/data/utils/boxes.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
+"""Helper functions for processing bounding box detections and annotations."""
+
```
Comment (samet-akcay) [inline anomalib/data/utils/boxes.py:2]: License is missing

File: anomalib/data/utils/boxes.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        List[Tensor]: A list of length B where each element is a tensor of shape (N, 4) containing the bounding box
+            coordinates of the objects in the masks in xyxy format.
+    """
+    masks = masks.view((-1, 1) + masks.shape[-2:])  # reshape to (B, 1, H, W)
+    masks = masks.float()
+
+    if masks.is_cuda:
+        batch_comps = connected_components_gpu(masks).squeeze(1)
+    else:
+        batch_comps = connected_components_cpu(masks).squeeze(1)
```
Comment (samet-akcay) [inline anomalib/data/utils/boxes.py:27]: A bit opinionated, but how about this? You don't need to accept this change.
```suggestion
    connected_components = connected_components_gpu if masks.is_cuda else connected_components_cpu
    batch_comps = connected_components(masks).squeeze(1)
```

File: anomalib/deploy/inferencers/openvino_inferencer.py [hunk 12]
```diff
@@ -159,7 +160,7 @@ def post_process(
         if "image_threshold" in meta_data:
             pred_label = pred_score >= meta_data["image_threshold"]
 
-        if task == "segmentation":
+        if task == TaskType.SEGMENTATION:
```
Comment (samet-akcay) [inline anomalib/deploy/inferencers/openvino_inferencer.py:163]: Do we have any plans on supporting detection in OpenVINO inference?

File: anomalib/models/components/base/anomaly_module.py [hunk 13]
```diff
@@ -82,6 +84,8 @@ def predict_step(self, batch: Any, batch_idx: int, _dataloader_idx: Optional[int
         outputs["pred_labels"] = outputs["pred_scores"] >= self.image_threshold.value
         if "anomaly_maps" in outputs.keys():
             outputs["pred_masks"] = outputs["anomaly_maps"] >= self.pixel_threshold.value
+            if "pred_boxes" not in outputs.keys():
+                outputs["pred_boxes"] = masks_to_boxes(outputs["pred_masks"])
```
Comment (samet-akcay) [inline anomalib/models/components/base/anomaly_module.py:88]: Is it possible to add `pred_boxes` to `outputs` only when the task type is `detection`? Or am I missing something?

File: anomalib/models/components/base/anomaly_module.py [hunk 14]
```diff
@@ -155,15 +159,28 @@ def _collect_outputs(image_metric, pixel_metric, outputs):
...
-
-    @staticmethod
-    def _outputs_to_cpu(output):
-        for key, value in output.items():
-            if isinstance(value, Tensor):
-                output[key] = value.cpu()
+        elif "pred_scores" not in outputs and "boxes_scores" in outputs:
+            # infer image score from bbox confidence scores
+            outputs["pred_scores"] = torch.stack([scores.max() for scores in outputs["boxes_scores"]])
+        if "pred_boxes" in outputs and "anomaly_maps" not in outputs:
```
Comment (samet-akcay) [inline anomalib/models/components/base/anomaly_module.py:169]: Can we have a new line above as it initially looks confusing to see an `if ` statement right after `elif`

File: anomalib/utils/cv/connected_components.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
+"""Connected component labeling."""
+
```
Comment (samet-akcay) [inline anomalib/utils/cv/connected_components.py:2]: License?

**Review by samet-akcay — COMMENTED** [review]
Took me some time to review :) Appreciate your time to finish the entire implementation :)

**Comment by jpcbertoldo** [comment]
> Using enum instead of string to configure task type.

I found a few str <--> enum conversion issues at some point when trying to use the cli with config files (somehow the string from the `.yaml` was not properly converted?).

So it made me wonder: are there tests covering this? (i.e. the different values of the enum work properly when being passed from the config file)

File: anomalib/utils/callbacks/min_max_normalization.py [hunk 16]
```diff
@@ -46,10 +47,14 @@ def on_validation_batch_end(
...
-        if "anomaly_maps" in outputs.keys():
+        if "anomaly_maps" in outputs:
             pl_module.normalization_metrics(outputs["anomaly_maps"])
-        else:
+        elif "boxes_scores" in outputs:
+            pl_module.normalization_metrics(torch.cat(outputs["boxes_scores"]))
+        elif "pred_scores" in outputs:
             pl_module.normalization_metrics(outputs["pred_scores"])
+        else:
+            raise ValueError("No values found for normalization, provide anomaly maps, bbox scores, or image scores")
```
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/utils/callbacks/min_max_normalization.py:57]: I agree that `image_score` would be more suitable, but I feel changing this is outside of the scope of this PR.

File: anomalib/data/__init__.py (hunk 1 above)
Comment (djdameln) [reply anomalib/data/__init__.py:17]: Yeah, I'm not super happy with the location of `TaskType` but couldn't think of a better place. Maybe somewhere in `anomalib/config` would work? @ashwinvaidya17 any ideas?

File: anomalib/data/base/datamodule.py (hunk 2 above)
Comment (djdameln) [reply anomalib/data/base/datamodule.py:25]: elem stands for 'element'. The naming was chosen to match the naming used in [torch's default collate function](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/collate.py#L115). A single element is taken from the batch to check if the batch is a list of dictionaries, and to retrieve the keys of the dictionary that need to be collated.

I'll add an inline comment to clarify this.

File: anomalib/data/base/video.py (hunk 5 above)
Comment (djdameln) [reply anomalib/data/base/video.py:88]: It turns out this was unnecessarily complicated and could be reduced to a single assignment.

File: anomalib/deploy/inferencers/openvino_inferencer.py (hunk 12 above)
Comment (djdameln) [reply anomalib/deploy/inferencers/openvino_inferencer.py:163]: Inference support will follow. I don't see any problems with adding OpenVINO support, but will need to check to confirm.

File: anomalib/models/components/base/anomaly_module.py (hunk 13 above)
Comment (djdameln) [reply anomalib/models/components/base/anomaly_module.py:88]: The problem is that we can't access task type information from within the lightning module. This is one of the reasons why we should consider switching to custom training loops.

File: anomalib/data/utils/boxes.py (hunk 11 above)
Comment (djdameln) [reply anomalib/data/utils/boxes.py:30]: mypy was causing some trouble after I committed this change, so I reverted it.

**Comment by djdameln** [comment]
> So it made me wonder: are there tests covering this? (i.e. the different values of the enum work properly when being passed from the config file)

I don't think the tests cover this explicitly at the moment. Which enum was causing problems at your end?

**Comment by jpcbertoldo** [comment]
> > So it made me wonder: are there tests covering this? (i.e. the different values of the enum work properly when being passed from the config file)
> 
> I don't think the tests cover this explicitly at the moment. Which enum was causing problems at your end?

Honestly I don't remember because it was blocking me to use the new cli so I dropped the ball.
Anyway, just something to keep in mind : )

**Comment by djdameln** [comment]
@ashwinvaidya17 @samet-akcay I've updated the PR based on your comments, and also added test cases and inference support. So it's ready for the next round of review.

File: anomalib/utils/callbacks/visualizer/visualizer_base.py [hunk 17]
```diff
@@ -37,8 +38,10 @@ def __init__(
         if mode not in ["full", "simple"]:
             raise ValueError(f"Unknown visualization mode: {mode}. Please choose one of ['full', 'simple']")
         self.mode = mode
-        if task not in ["classification", "segmentation"]:
-            raise ValueError(f"Unknown task type: {mode}. Please choose one of ['classification', 'segmentation']")
+        if task not in [TaskType.CLASSIFICATION, TaskType.DETECTION, TaskType.SEGMENTATION]:
+            raise ValueError(
+                f"Unknown task type: {mode}. Please choose one of ['classification', 'detection', 'segmentation']"
```
Comment (ashwinvaidya17) [inline anomalib/utils/callbacks/visualizer/visualizer_base.py:43]: Not really necessary here but I wonder if `f"Unknown task type: {mode}. Please choose one of {TaskType.__members__.values()}` will work here. But will not make it a very descriptive docstring for the programmers.

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks for the efforts!

**Review by samet-akcay — APPROVED** [review]
LGTM, thanks!
